mitiq/shadows/shadows_utils.py

```diff
@@ -0,0 +1,157 @@
+    """
+    M = len(observables)
+    K = 2 * np.log(2 * M / failure_rate)
+
+    N = (
+        34
+        * max(local_clifford_shadow_norm(o) for o in observables)
+        / error**2
+    )
+    return int(np.ceil(N * K)), int(K)
+
+
+def fidelity(
+    state_vector: NDArray[np.complex64],
+    rho: NDArray[np.complex64],
+) -> float:
+    """
+    Calculate the fidelity between a state vector and a density matrix.
+    Args:
+        state_vector: The vector whose norm we want to calculate.
+        rho: The operator whose norm we want to calculate.
+
+    Returns:
+        Scalar corresponding to the fidelity.
+    """
+    return np.reshape(state_vector.conj().T @ rho @ state_vector, -1).real[0]
+
+
+def transform_to_cirq_paulistring(
+    pauli_str: Union[str, mitiq.PauliString, cirq.PauliString[Any]]
+) -> Tuple[float, cirq.PauliString[Any]]:
+    """Transforms mitiq.PauliString, cirq.PauliString, cirq.PauliSum or string
+    to a cirq.PauliString class.
+
+    Args:
+      pauli_str: A mitiq.PauliString, cirq.PauliString, cirq.PauliSum
+        or string.
+
+    Returns:
+      A tuple containing the coefficient and a cirq.PauliString.
+    """
+    if isinstance(pauli_str, cirq.PauliString):
+        # Extract coefficient and return cirq.PauliString without coefficient
+        coeff = np.real(pauli_str.coefficient)  # type: ignore
+        new_pauli_str: cirq.PauliString[Any] = cirq.PauliString(
+            {qubit: op for qubit, op in pauli_str.items()}
+        )
+        return coeff, new_pauli_str
+
+    elif isinstance(pauli_str, mitiq.PauliString):
+        qubit_pauli_map = {
+            cirq.LineQubit(i): gate
+            for i, gate in enumerate(pauli_str._pauli.values())
+        }
+        return np.real(pauli_str.coeff), cirq.PauliString(
+            qubit_pauli_map=qubit_pauli_map
+        )
+    elif isinstance(pauli_str, str):
+        # Extract the coefficient from the start of the string
+        matched_results = re.match(
+            r"([+-]?(\d+(\.\d*)?|\.\d+))?(.*)", pauli_str
+        )
+        if matched_results is None:
+            raise ValueError("pauli_str should be of format '0.5IXXY'")
+        coeff_str, _, _, pauli_str = matched_results.groups()
+        coefficient = float(coeff_str) if coeff_str else 1.0
+
+        # Create a mapping for Pauli gates
+        pauli_map = {"X": cirq.X, "Y": cirq.Y, "Z": cirq.Z}
+        # Create a dictionary where the keys are qubits and the values are
+        # corresponding Pauli operations
+        qubit_pauli_map = {}
+        for i, pauli_name in enumerate(pauli_str):  # X, Y, Z, I
+            if pauli_name == "I":
+                continue
+            qubit_pauli_map[cirq.LineQubit(i)] = pauli_map[pauli_name]
+
+        return coefficient, cirq.PauliString(qubit_pauli_map=qubit_pauli_map)
+
+    else:
+        raise ValueError(
+            "pauli_str must be cirq.PauliString, mitiq.PauliString, "
+            "cirq.PauliSum or string."
+        )
```

Comment:
Raising a `TypeError` would be more appropriate here.

Reply:
have deleted this function.